myia/ir/visualization/graph_printer.py

```diff
@@ -0,0 +1,277 @@
+    ).read()
+
+    def __init__(
+        self,
+        graph: Graph,
+        *,
+        on_node=None,
+        show_constants=True,
+        link_fn_graphs=True,
+        link_inp_graphs=True,
+    ):
+        """Initialize.
+
+        :param graph: graph to visualize
+        :param on_node: optional function to call on printed node when clicked
+        :param show_constants: if True, display constant non-graph nodes.
+            Constant graphs are always displayed.
+        :param link_fn_graphs: if True, display edges from apply FNs to constant graphs
+        :param link_inp_graphs: if True, display edges from apply inputs to constant graphs
+        """
+        self.graphs, self.nodes, self.edges = self.collect_myia_elements(
+            graph, show_constants, link_fn_graphs, link_inp_graphs
+        )
+        self._on_node = on_node
+        self._lbl = _NodeCache()
+
+    def on_node(self, data):  # pragma: no cover
+        """Callback on given data when a node is clicked.
+
+        Not called in tests. Needs snektalk to be tested.
+
+        :param data: dictionary representing clicked node.
+            Contains at least "id" and "label" fields
+        """
+        if not self._on_node:
+            return
+        return self._on_node(data)
+
+    @classmethod
+    def __hrepr_resources__(cls, H):
+        return [
+            H.javascript(
+                src="https://cdnjs.cloudflare.com/ajax/libs/cytoscape/3.17.0/cytoscape.min.js",
+                export="cytoscape",
+            ),
+            H.javascript(
+                src="https://cdn.rawgit.com/cytoscape/cytoscape.js-dagre/1.5.0/cytoscape-dagre.js",
+                export="cytoscape-dagre",
+            ),
+            H.javascript(
+                src="https://cdn.rawgit.com/cpettitt/dagre/v0.7.4/dist/dagre.js",
+                export="dagre",
+            ),
+            H.javascript(
+                """
+                // Load dagre layout
+                cydagre(cytoscape, dagre);
+
+                function make_graph(element, options) {
+                    options.container = element;
+                    this.cy = cytoscape(options);
+                    if (options.on_node) {
+                        this.cy.on('click', 'node', function(evt){
+                            options.on_node(evt.target.data());
+                        });
+                    }
+                }
+                """,
+                require={
+                    "cytoscape": "cytoscape",
+                    "dagre": "dagre",
+                    "cytoscape-dagre": "cydagre",
+                },
+                export="make_graph",
+            ),
+        ]
+
+    def __hrepr__(self, H, hrepr):
+        # Generate identifiers to make hrepr output deterministic.
+        elements = self.graphs + self.nodes
+        identifiers = {
+            element: indice for indice, element in enumerate(elements)
```

Comment:
Probably better to convert them all to strings upfront.

```suggestion
            element: str(index) for index, element in enumerate(elements)
```

Reply:
Ok !